rtt/src/rtt.rs

```diff
@@ -0,0 +1,267 @@
+use probe_rs::{config::MemoryRegion, MemoryInterface, Session};
+use scroll::{Pread, LE};
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::ops::Range;
+use std::sync::{Arc, Mutex};
+
+use crate::channel::*;
+use crate::{Channels, Error};
+
+/// The RTT interface.
+///
+/// Use [`Rtt::attach`] to attach to a probe-rs `Core` and detect channels.
+#[derive(Debug)]
+pub struct Rtt {
+    ptr: u32,
+    up_channels: Channels<UpChannel>,
+    down_channels: Channels<DownChannel>,
+}
+
+// Rtt must follow this data layout when reading/writing memory in order to be compatible with the
+// official RTT implementation.
+//
+// struct ControlBlock {
+//     char id[16]; // Used to find/validate the control block.
+//     // Maximum number of up (target to host) channels in following array
+//     unsigned int max_up_channels;
+//     // Maximum number of down (host to target) channels in following array.
+//     unsigned int max_down_channels;
+//     RttChannel up_channels[max_up_channels]; // Array of up (target to host) channels.
+//     RttChannel down_channels[max_down_channels]; // array of down (host to target) channels.
+// }
+
+impl Rtt {
+    const RTT_ID: [u8; 16] = *b"SEGGER RTT\0\0\0\0\0\0";
+
+    // Minimum size of the ControlBlock struct in target memory in bytes with empty arrays
+    const MIN_SIZE: usize = Self::O_CHANNEL_ARRAYS;
+
+    // Offsets of fields in target memory in bytes
+    const O_ID: usize = 0;
+    const O_MAX_UP_CHANNELS: usize = 16;
+    const O_MAX_DOWN_CHANNELS: usize = 20;
+    const O_CHANNEL_ARRAYS: usize = 24;
+
+    fn from(
+        session: Arc<Mutex<Session>>,
+        memory_map: &[MemoryRegion],
+        // Pointer from which to scan
+        ptr: u32,
+        // Memory contents read in advance, starting from ptr
+        mem_in: Option<&[u8]>,
+    ) -> Result<Option<Rtt>, Error> {
+        let mut mem = match mem_in {
+            Some(mem) => Cow::Borrowed(mem),
+            None => {
+                // If memory wasn't passed in, read the minimum header size
+                let mut mem = vec![0u8; Self::MIN_SIZE];
+                let mut lock = session.lock().unwrap();
+                let mut core = lock.core(0)?;
+                core.read_8(ptr, &mut mem)?;
+                Cow::Owned(mem)
+            }
+        };
+
+        // Validate that the control block starts with the ID bytes
+        let rtt_id = &mem[Self::O_ID..(Self::O_ID + Self::RTT_ID.len())];
+        if rtt_id != Self::RTT_ID {
+            log::trace!(
+                "Expected control block to start with RTT ID. Got instead: {:?}",
+                rtt_id
+            );
+
+            return Ok(None);
+        }
+
+        let max_up_channels = mem.pread_with::<u32>(Self::O_MAX_UP_CHANNELS, LE).unwrap() as usize;
+        let max_down_channels = mem
+            .pread_with::<u32>(Self::O_MAX_DOWN_CHANNELS, LE)
+            .unwrap() as usize;
+
+        // *Very* conservative sanity check, most people
```

Comment:
Most people?

Reply:
I guess most people wont use more channels. I did not write this code but ported it - with authors approval.

Reply:
Ahh, I just came across it and assumed you forgot to finish the sentence